cpp/src/gandiva/projector.cc

```diff
@@ -175,6 +175,12 @@ Status Projector::AllocArrayData(const DataTypePtr& type, int64_t num_records,
   astatus = arrow::AllocateBuffer(pool, data_len, &data);
   ARROW_RETURN_NOT_OK(astatus);
 
+  // Valgrind detects unitialized memory at byte level. Boolean types use bits
+  // and can leave buffer memory uninitialized in the last byte.
+  if (type->id() == arrow::Type::BOOL) {
+    data->mutable_data()[data_len - 1] = 0;
+  }
```

Comment:
Is it really being left uninitialized? 

FWIW we zero the memory in `arrow::BooleanBuilder` because valgrind doesn't like in-place modifications of uninitialized bytes

https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_primitive.cc#L167

Reply:
gandiva doesn't use the builders - it allocates the buffers directly in cpp code (for the batch), and update the buffers in IR code. using builders is tricky since they expect the updates also to happen through the builder APIs (eg. for tracking length).

> Is it really being left uninitialized?

gandiva only updates the relevant bits. eg. for a projector with expression "a < b" having a batch of 6 elements, gandiva will update 6 bits in the output boolean vector (to either 0 or 1 depending on the values of a and b). The remaining 2 bits are left uninitialized.

Reply:
I wasn't suggesting that you use the builders, just noting that we've also experienced valgrind issues with boolean arrays

Uninitialized bits are not an issue. I was curious why a _whole byte_ is uninitialized

Reply:
I'm still curious about the answer to this

Reply:
I think what it means is that avoid Valgrind errors with uninitialized bits.

Reply:
Maybe, valgrind does require us to zero out the entire bitmap. @shyambits2004, can you please check if we do have a unit test that projects more than 8 elements?

Reply:
@pitrou agreed, we zero out all of our bytes in BooleanBuilder for example. I'm trying to understand why the _last byte_. That is what seems weird to me -- per @pravindra it may be that the testing is not comprehensive enough

Reply:
I'm not sure, but Valgrind is able to detect individual uninitialized bits. So only the trailing bits in the last byte would be a problem.

It also depends which exact operation is used for setting or clearing the bits.

Reply:
I wasn't aware that valgrind had bit-level precision (http://valgrind.org/docs/memcheck2005.pdf) so this is probably it.

Reply:
@wesm, you are right. valgrind complained when I modified the test to have 12 output elements. gandiva uses arrow::util::SetBitTo() to update bitmaps, and there's a comment in the function that it confuses valgrind.

I've opened ARROW-4115 for this.